compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+from collections.abc import Sequence
+
+from xdsl.context import MLContext
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.dialects.builtin import IndexType
+from xdsl.ir import Block, Operation, OpResult, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+
+
+class MoveMemrefAllocations(RewritePattern):
+    """
+    This class represents a rewrite pattern for moving memref allocations outside
+    a double for-loop. This is possible when each loop a new memref-space is allocated with
+    identical sizes and the space is not used outside the loop.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, alloc_op: memref.Alloc, rewriter: PatternRewriter):
+        def find_parent_for_loop(op: Operation) -> scf.For:
+            """
+            Find the parent for-loop of the operation.
+            If no for-loop is found, return None.
+            """
+            op = op.parent_op()
+            while not isinstance(op, scf.For):
+                if op.parent_op() is None:
+                    return None
+                op = op.parent_op()
+            return op
+
+        def before_loop(op: Operation) -> bool:
+            """
+            Check if the operation is defined before the loop.
+            """
+            return find_parent_for_loop(op) is not find_parent_for_loop(alloc_op)
+
+        def is_in_loop(op: Operation) -> bool:
+            """
+            Check if the operation is inside a loop.
+            """
+            return find_parent_for_loop(op) is not None
+
+        def can_move_dim(dim: memref.Dim) -> bool:
+            """
+            Check if the Dim operation can be moved outside the loop.
+            """
+            if before_loop(dim):
+                return True
+            subview = dim.source.owner
+            if isinstance(subview, Block):
+                # This happens when the dim is called on an input argument
+                return True
+            assert isinstance(subview, memref.Subview)
+            index = int(dim.index.owner.value.value.data)
```

Comment:
There is a third case we should handle here:
If there are two matmuls after each other, like
```
A = B * C
F = A * G
```
For the calculation of `F = A * G`, we need to allocate space for A. A is not the result of a subview, but instead, the result of a previous `alloc` operation.